packages/dds/tree/src/feature-libraries/default-schema/schemaChecker.ts

```diff
@@ -0,0 +1,153 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { assert, unreachableCase } from "@fluidframework/core-utils/internal";
+import {
+	MapTree,
+	StoredSchemaCollection,
+	TreeFieldStoredSchema,
+	LeafNodeStoredSchema,
+	ObjectNodeStoredSchema,
+	MapNodeStoredSchema,
+	Multiplicity,
+} from "../../core/index.js";
+import { FullSchemaPolicy } from "../modular-schema/index.js";
+import { allowsValue } from "../valueUtilities.js";
+
+export const enum SchemaValidationErrors {
+	NoError,
+	Field_KindNotInSchemaPolicy,
+	Field_IncorrectMultiplicity,
+	Field_NodeTypeNotAllowed,
+	LeafNode_MissingValue,
+	LeafNode_InvalidValue,
+	LeafNode_FieldsNotAllowed,
+	ObjectNode_FieldCountMismatch,
+	ObjectNode_FieldNotInSchema,
+	Node_MissingSchema,
+	UnknownError,
+}
+
+export function isNodeInSchema(
+	node: MapTree,
+	nodeSchemaCollection: StoredSchemaCollection,
+	schemaPolicy: FullSchemaPolicy,
+): SchemaValidationErrors {
+	// Validate the schema declared by the node exists
+	const schema = nodeSchemaCollection.nodeSchema.get(node.type);
+	if (schema === undefined) {
+		return SchemaValidationErrors.Node_MissingSchema;
+	}
+
+	// Validate the node is well formed according to its schema
+
+	if (schema instanceof LeafNodeStoredSchema) {
+		if (node.value === undefined) {
+			return SchemaValidationErrors.LeafNode_MissingValue;
+		}
```

Comment:
If desired, you could remove this case, since allowsValue catches it below: we can just consider undefined an invalid value for all leaves that way.

Reply:
Good catch, I'll remove it to simplify.